File: elixir/lib/advent_of_code/solution/year_2023/day_01.ex
defmodule AdventOfCode.Solution.Year2023.Day01 do
  def part1(input) do
    input
    |> String.split("\n", trim: true)
    |> Enum.map(&line_to_int/1)
    |> Enum.sum()
  end

  def part2(input) do
    input
    |> String.split("\n", trim: true)
    |> Enum.map(&line_to_part2_line/1)
    |> Enum.map(&line_to_int/1)
    |> Enum.sum()
  end

  defp line_to_int(line) do
    integers =
      line
      |> String.graphemes()
      |> Enum.map(&Integer.parse/1)
      |> Enum.filter(&is_tuple/1)
      |> Enum.map(&elem(&1, 0))

    Integer.undigits([List.first(integers), List.last(integers)])
  end

  defp line_to_part2_line(line) do
    line
    |> String.replace("one", "o1e")
    |> String.replace("two", "t2o")
    |> String.replace("three", "t3e")
    |> String.replace("four", "f4r")
    |> String.replace("five", "f5e")
    |> String.replace("six", "s6x")
    |> String.replace("seven", "s7n")
    |> String.replace("eight", "e8t")
    |> String.replace("nine", "n9e")
  end
end

File: python/solution/year_2023/day_01.py
def part1(input):
    lines = input.split("\n")
    return part1_from_lines(lines)   

def part1_from_lines(lines):
    first_digits = [first_digit(line) for line in lines]
    last_digits = [last_digit(line) for line in lines]
    nums = zip(first_digits, last_digits)
    return sum([first*10+last for first, last in nums])

def part2(input):
    lines = input.split("\n")
    new_lines = [part2_line_to_part1_line(line) for line in lines]
    return part1_from_lines(new_lines)           

def first_digit(line):
    for char in line:
        if char.isdigit():
            return int(char)
    raise ValueError("No digit found in line")

def last_digit(line):
    return first_digit(line[::-1])

word2num = {
    "one": "o1e",
    "two": "t2o",
    "three": "t3e",
    "four": "f4r",
    "five": "f5v",
    "six": "s6x",
    "seven": "s7n",
    "eight": "e8t",
    "nine": "n9e",
}

def part2_line_to_part1_line(line):
    for word, num in word2num.items():
        line = line.replace(word, num)
    return line